### ISSUE TITLE
RuntimeError: cannot reuse already awaited coroutine

### ISSUE THREAD

**Issue by kinow** [cylc/cylc-uiserver#108]
**Describe the bug**

From: https://github.com/cylc/cylc-uiserver/pull/107#issuecomment-576061827

The following error happens in runtime:

```bash
od IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOMainLoop object at 0x7fb41c96e7b8>>, <Task finished coro=<TornadoSubscriptionServer.handle() done, defined at /home/kinow/Development/python/workspace/cylc-uiserver/cylc/uiserver/websockets/tornado.py:79> exception=RuntimeError('cannot reuse already awaited coroutine')>)
Traceback (most recent call last):
  File "/home/kinow/Development/python/workspace/cylc-uiserver/venv/lib/python3.7/site-packages/tornado/ioloop.py", line 743, in _run_callback
    ret = callback()
  File "/home/kinow/Development/python/workspace/cylc-uiserver/venv/lib/python3.7/site-packages/tornado/ioloop.py", line 767, in _discard_future_result
    future.result()
RuntimeError: cannot reuse already awaited coroutine
```

**Release version(s) and/or repository branch(es) affected?**

`master`

**Steps to reproduce the bug**

Run some workflows for a while, and switch views in the UI. Eventually this errors pops in the console output.

**Expected behavior**

No errors.

**Screenshots**

**Additional context**

@dwsutherland commented:

>Might be if the async gen is not being exited correctly, I think it might need the yield within a try: & await/other within a finally: .. doesn’t need to be addressed here.

**Pull requests welcome!**
This is an Open Source project - please consider contributing a bug fix
yourself (please read `CONTRIBUTING.md` before starting any work though).

**Comment by kinow** [comment]
This looks like an normal error that happens when the coroutine that is created by the WebSocket handler

https://github.com/cylc/cylc-uiserver/blob/1d35ff47ab843a1ff8823c95094e967c29ab248d/cylc/uiserver/handlers.py#L127-L129

To quickly reproduce it, just start up the whole Cylc system, then go to the UI, wait for it to load. Now fire a few refreshes (e.g. `CTRL + SHIFT + R` on Firefox Linux a couple times). The callback will be created, then discarded, and another one will be created, and eventually be used.

Looks like when the callback is added, but discarded, it is awaited twice. My guess is that the first time is the legit call, which is interrupted. Then the second is the normal Tornado code that tries to prevent exceptions not being logged.

![image](https://user-images.githubusercontent.com/304786/76029323-1b3f7500-5f99-11ea-9ae6-4acfc2f47c0f.png)

(this error is logged on that `except Exception` line that prints "Exception in callback ...")

**Comment by kinow** [comment]
There are errors that may happen when using WebSockets and can be ignored (e.g. https://github.com/tornadoweb/tornado/issues/2763#issuecomment-557815906). I guess this error is in the same category. Couldn't find anything we could do to prevent it from being logged without adding a `except RuntimeError` that looks quite dangerous I guess.

**Comment by oliver-sanders** [comment]
Oh dear, this one keeps getting bumped, has anyone managed to reproduce this recently?

**Comment by kinow** [comment]
> Oh dear, this one keeps getting bumped, has anyone managed to reproduce this recently?

I haven't seen this one in a long time. @dwsutherland have you seen this bug?

@oliver-sanders if @dwsutherland confirms he hasn't seen it, tomorrow I will try to reproduce as per issue description above (refreshing the UI), and if I don't see, we can close as cannot reproduce. It might have been fixed by some of the many changes since it was reported :man_shrugging:

**Comment by kinow** [comment]
@dwsutherland confirmed in Element he hasn't seen it.

>no I haven't, but I'd suggest getting the current batch of fixes in so we can rule them out
maybe reproduce by hitting mutations really fast?

Let's keep it open just until the release. Then we can try firing a few mutations in parallel (craft a command line call and `xargs` or `parallel` it?), and refreshing the UI a few times too. If no one is able to reproduce it, then we close it instead of bumping it to the next milestone. WDYT?

**Comment by kinow** [comment]
I tried to reproduce another issue, and then decided to try this one too. Just for the record, no success in reproducing it after ~5 minutes trying. Switched views (click on the workflow in GScan, then on another one, which stops/starts subscriptions), paused/played. Triggered some tasks. The error never happened :+1:

**Comment by oliver-sanders** [comment]
I've not seen this for a while either.

The event loop may have changed somewhat since we moved to JupyterServer, tentatively closing now unless we hear of any other reports.
